teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,431 @@
+       which could become an interface for "git remotes"
+     * Potential groupings for the git-clone task stand out as well, for example the proxy configuration
+     * On that note, since the git-rebase task is using git and accessing remote repos, it probably needs the same
+       proxy configuration, so what they probably both need is some kind of interface for what values to provide when
+       accessing a remote git repo
+   Other examples include [okra-deploy](https://github.com/tektoncd/catalog/tree/main/task/orka-deploy/0.1) which is using
+   param name prefixes to group related params (e.g. `ssh-`, `okra-vm`, `okra-token`).
+
+## Requirements
+
+* Must be possible to programmatically determine the structure of the object
+* Must be possible for a result object to be empty
+* Must be possible to use the object results of one Task in a Pipeline as the
+  param of another Task in the pipeline which has a object param
+* Must be possible to use one specific value in a object result of one Task
+  in a Pipeline as the param of another Task in the pipeline which has a string param
+  * If there is no value for the specified key, the Pipeline execution should fail
+    (or we may be able to use [TEP-0048](https://github.com/tektoncd/community/pull/240)
+    to specify a default value to use in that case instead)
+
+## Proposal
+
+* We would add support for object types for results and params, in addition to the existing string and array support
+  * Initially we would only support string keys, eventually we can expand this to all values (string, array, object)
+    (note that this is the case for arrays as well, we don't yet support arrays of arrays)
+* We would use [json object schema syntax](https://github.com/tektoncd/pipeline/issues/1393#issuecomment-558833526)
+  to express the object structure
+* To support object results, we would support writing results to a `.json` file, as described in
+  [TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477)
+
+This feature would be considered alpha and would be
+[gated by the alpha flag](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#alpha-features).
+
+### Notes/Caveats
+
+[See also TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477) for notes and caveats
+specific to supporting json results.
+
+* What if a Pipeline tries to use a object key that doesn't exist in a param or result?
+  * Some of this could be caught at creation time, i.e. pipeline params would declare the keys each object param will
+    contain, and we could validate at creation time that those keys exist
+  * For invalid uses that can only be caught at runtime, the PipelineRun would fail
+* What if a Task tries to write more keys to a result than the result declares it contains?
+  * The TaskRun would fail (alternatively we could decide to ignore the extra keys)
```

Comment:
I wonder why would we want them to fail ? allowing "extra" *keys* means it would be more easy to integrate existing tools into tasks, or use a similar tool (that generate the same output) for different task with different results.

Reply:
Good point, I'll flip this around!

Reply:
ah and actually this lines up with [the `additionalProperties` config in json objects](https://json-schema.org/understanding-json-schema/reference/object.html#additional-properties) - which looks like it's true by default, i.e. allowing the inclusion of extra fields